src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSBwaAligner.java

```diff
@@ -0,0 +1,100 @@
+        this.numThreads = bwaArgs.bwaThreads;
+        this.maxAltHits = bwaArgs.maxAlternateHits;
+        this.scoreThreshold = bwaArgs.scoreThreshold;
+        this.pairedAlignment = pairedAlignment;
+    }
+
+    public Iterator<GATKRead> apply(final Iterator<GATKRead> itr, final SAMFileHeader header) {
+        //Create aligner and set options
+        final BwaMemAligner aligner = new BwaMemAligner(bwaIndex);
+        if (pairedAlignment) {
+            aligner.alignPairs();
+        }
+        aligner.setMaxXAHitsAltOption(maxAltHits);
+        aligner.setMaxXAHitsOption(maxAltHits);
+        aligner.setMinSeedLengthOption(minSeedLength);
+        aligner.setOutputScoreThresholdOption(scoreThreshold);
+        aligner.setNThreadsOption(numThreads);
+
+        //Get list of reads on the partition
+        final List<GATKRead> reads = new ArrayList<>();
+        while (itr.hasNext()) {
+            reads.add(itr.next());
+        }
+        final int numReads = reads.size();
+        if (pairedAlignment && numReads % 2 != 0) {
+            throw new UserException.BadInput("Expected paired reads but there are an odd number");
+        }
+
+        //Align read sequences
+        final List<Tuple2<GATKRead, List<BwaMemAlignment>>> results;
+        if (numReads == 0) {
+            results = Collections.emptyList();
+        } else {
+            final List<byte[]> seqs = reads.stream().map(GATKRead::getBases).collect(Collectors.toList());
+            final List<List<BwaMemAlignment>> alignments = aligner.alignSeqs(seqs);
+            results = new ArrayList<>(numReads);
+            for (int i = 0; i < numReads; i++) {
+                results.add(new Tuple2<>(reads.get(i), alignments.get(i)));
+            }
+        }
+
+        //Map alignments back to the reads
+        final List<String> refNames = bwaIndex.getReferenceContigNames();
+        return Utils.stream(results).map(item -> {
+            final GATKRead read = item._1;
+            SAMRecord rec = read.convertToSAMRecord(header);
+            final List<BwaMemAlignment> alignmentList = item._2;
+            for (final BwaMemAlignment alignment : alignmentList) {
+                //Only get primary alignments
+                if (SAMFlag.NOT_PRIMARY_ALIGNMENT.isUnset(alignment.getSamFlag())
+                        && SAMFlag.SUPPLEMENTARY_ALIGNMENT.isUnset(alignment.getSamFlag())) {
+                    if (rec.getReadUnmappedFlag()) {
+                        //Record is currently unmapped, so apply first alignment
+                        rec = BwaMemAlignmentUtils.applyAlignment(rec.getReadName(), rec.getReadBases(), rec.getBaseQualities(),
+                                rec.getReadGroup().getId(), alignment, refNames, header, false, false);
+                    } else if (alignment.getRefId() != -1) {
+                        //Record is currently mapped, but if the alignment is not empty then add as an alternate alignment
```

Comment:
If the original read is already mapped (ie in the input bam), it's possible the aligner doesn't come up with any alignments in which case it returns an alignment with refId -1. 

This isn't likely in the current tool implementation because it checks that the input BAM header doesn't have any sequences, but could be in future versions if we want to be able to align to multiple reference files in the pipeline. 

I added a comment to explain.